Move LayerBuffer cache to the compositor

mrobinson opened the pull request:
Now that NativeDisplay can be shared between the compositor and the
paint task, we can move the LayerBuffer cache to the compositor. This
allows surfaces to be potentially reused between different paint tasks
and will eventually allow OpenGL contexts to be preserved between
instances of GL rasterization.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/6530)

<!-- Reviewable:end -->

mrobinson committed:
Move LayerBuffer cache to the compositor

Now that NativeDisplay can be shared between the compositor and the
paint task, we can move the LayerBuffer cache to the compositor. This
allows surfaces to be potentially reused between different paint tasks
and will eventually allow OpenGL contexts to be preserved between
instances of GL rasterization.

mrobinson commented:
@gw r?

mrobinson commented:
@pcwalton You might also want to review this since I believe that you wrote the BufferMap originally.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #6551) made this pull request unmergeable. Please resolve the merge conflicts.

Ms2ger commented:
@glennw r?

pcwalton commented on components/compositing/compositor.rs:
```diff
@@ -290,6 +295,7 @@ impl<Window: WindowMethods> IOCompositor<Window> {
             last_composite_time: 0,
             has_seen_quit_event: false,
             ready_to_save_state: ReadyState::Unknown,
+            buffer_map: BufferMap::new(10000000),
```
Pull this out into a `const`.

pcwalton commented on components/compositing/compositor.rs:
```diff
@@ -1141,7 +1153,32 @@ impl<Window: WindowMethods> IOCompositor<Window> {
+                return;
+            }
+
+            match self.buffer_map.find(buffer_request.screen_rect.size) {
```
Use `if let` to reduce indentation

pcwalton commented on components/gfx/paint_task.rs:
```diff
@@ -754,11 +623,59 @@ impl WorkerThread {
+            debug!("painting worker thread uploading to native surface {}",
+                   layer_buffer.native_surface.get_id());
+        });
+        return layer_buffer
```
nit: no need for `return` keyword here

pcwalton commented on components/gfx/paint_task.rs:
```diff
@@ -754,11 +623,59 @@ impl WorkerThread {
+                                                draw_target: DrawTarget,
+                                                scale: f32)
+                                                -> Box<LayerBuffer> {
+        let mut layer_buffer = match tile.layer_buffer {
```
nit: if you use `tile.layer_buffer.unwrap_or_else(|| { ... })` you can save a level of indentation

pcwalton commented:
Looks good to me other than those nits.

mrobinson commented:
@pcwalton Thanks for the review. I think I have addressed all the nits in the latest branch. One thing is that instead of `tile.layer_buffer.unwrap_or_else(|| { ... })`, I had to use `tile.layer_buffer.take().unwrap_or_else(|| { ... })` to avoid a double-borrow.

pcwalton commented:
@bors-servo: r+

Looks good!

bors-servo commented:
:pushpin: Commit e0a8c91 has been approved by `pcwalton`

<!-- @bors-servo r=pcwalton e0a8c91eb3728dfdda40f884f426561e04b06d32 -->

bors-servo commented:
:hourglass: Testing commit e0a8c91 with merge 0855fe0...

bors-servo commented:
:broken_heart: Test failed - [mac2](http://build.servo.org/builders/mac2/builds/2468)

jdm commented:
Missing file?
```
/Users/servo/buildbot/slave/mac2/build/components/compositing/lib.rs:47:5: 47:15 error: file not found for module `buffer_map`
/Users/servo/buildbot/slave/mac2/build/components/compositing/lib.rs:47 mod buffer_map;
                                                                            ^~~~~~~~~~
```

mrobinson commented:
Ugh, yes. I forgot to check in buffer_map.rs after it was moved to compositing. I've uploaded a new patch now.

mrobinson commented:
@bors-servo r=pcwalton

bors-servo commented:
:pushpin: Commit e115c3d has been approved by `pcwalton`

<!-- @bors-servo r=pcwalton e115c3d3c4996dd9014f0314be3f03f6ba106ee9 -->

bors-servo commented:
:hourglass: Testing commit e115c3d with merge 10b0d8c...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/2243), [gonk](http://build.servo.org/builders/gonk/builds/1397), [linux1](http://build.servo.org/builders/linux1/builds/2243), [linux2](http://build.servo.org/builders/linux2/builds/2248), [linux3](http://build.servo.org/builders/linux3/builds/400), [mac1](http://build.servo.org/builders/mac1/builds/2469), [mac2](http://build.servo.org/builders/mac2/builds/2472), [mac3](http://build.servo.org/builders/mac3/builds/421)

nnethercote commented:
Oh, you removed the memory reporting that was being done on BufferMap and didn't replace it with anything. That's... frustrating.

metajack commented:
@mrobinson Can you do a followup PR to put this reporting back?

mrobinson commented:
@metajack @nnethercote Yes, I plan to restore the memory reporting, but because the allocation has moved to the compositor task it makes sense to do it in a separate PR. I didn't expect to land this change so quickly, but because of timing issues with the rustup, it had to land today at the last minute.

mrobinson commented:
Hopefully tomorrow is not too late.
